### DISCUSSION TITLE
Next/script fires an onLoad and onReady event while script is onError after navigating

### DISCUSSION THREAD

**Discussion by RamziSayagh** [vercel/next.js#45528] [Help]
Hi everyone,

I wanted to use the onReady event provided by the Script component, but I realised that after navigating, by using Next/Link and going back to the page that loads the script, first the onLoad event is fired then the onReady.
The script was initially on error, and still not found.

```javascript
<Script
    src='scriptnotfound' 
    onLoad={() => {console.log('SCRIPT LOADED')}}
    onReady={() => {console.log('SCRIPT READY')}}
    onError={() => {console.log('SCRIPT ERROR')}}
/>
```

<img width="612" alt="image" src="https://user-images.githubusercontent.com/568853/216400757-34006a92-9dee-4c7e-af5e-6633d44c7c67.png">

Is that behavior wanted ?

**Comment by icyJoseph** [comment]
It looks like when the src is a so called "remote script", and requires "loading", a promise is created, which either resolves on load or rejects on error. This promise though, is directly saved to a cache.

```js
    ScriptCache.set(src, loadPromise) // this happens before loadPromise resolves/rejects
```

During this stage if the promise resolves, it is saved to a Set cache and onReady is called, otherwise, onError is called. 

Later on, during mount, the Set cache is queried for this src, and if found onReady is called.

I think the problem is that since the promise is added to a Map, called `ScriptCache`, and it is not removed on error, on the next render pass, when we hit this section:

```js
  // Contents of this script are already loading/loaded
  if (ScriptCache.has(src)) {
    LoadCache.add(cacheKey)
    // It is possible that multiple `next/script` components all have same "src", but has different "onLoad"
    // This is to make sure the same remote script will only load once, but "onLoad" are executed in order
    ScriptCache.get(src).then(onLoad, onError)
    return
```

I think there's a subtle bug on that `ScriptCache.get` line, consider:

```js
const throws = new Promise((resolve, reject) => {
  reject(new Error("Always reject"));
}).catch(() => console.log("rejected"));

async function main() {
  const promise = throws;

  promise.then(
    () => console.log("left"),
    () => console.log("right")
  );
}

main();
```

That'll log, `left`, which leads me to believe that the `onError` branch in `ScriptCache.get` is not doing its job 😅 

I'll try it out with the actual next.js source code later today.

**Reply by RamziSayagh** [reply]
Hi @icyJoseph, thanx for your reply !

You pointed out the thing, the promise rejected state is later resolved so we can not rely on the onLoad/onReady callback of the Script component.

I tried out those Next versions, unfortunately with the same subtle bug :
- 12.3.4
- 13.1.6
- 13.1.7-canary.4

**Reply by icyJoseph** [reply]
I made a PR https://github.com/vercel/next.js/pull/45558 hopefully that was it. It worked fine when I tested locally.
